src/Toml/BiMap.hs

```diff
@@ -138,158 +170,169 @@ _Just = prism Just id
 -- | Creates prism for 'AnyValue'.
 mkAnyValueBiMap :: (forall t . Value t -> Maybe a)
                 -> (a -> Value tag)
-                -> BiMap a AnyValue
+                -> TomlBiMap a AnyValue
 mkAnyValueBiMap matchValue toValue =
-    BiMap (Just . AnyValue . toValue) (\(AnyValue value) -> matchValue value)
+    BiMap (Right . AnyValue . toValue) (\(AnyValue value) -> maybeToBiEither $ matchValue value)
 
 -- | Creates bimap for 'Text' to 'AnyValue' with custom functions
-_TextBy :: (a -> Text) -> (Text -> Maybe a) -> BiMap a AnyValue
+_TextBy :: (a -> Text) -> (Text -> Maybe a) -> TomlBiMap a AnyValue
 _TextBy toText parseText =
   mkAnyValueBiMap (matchText >=> parseText) (Text . toText)
```

Comment:
This is a bit difficult case. `TextBy` actually can fail for multiple reasons:

1. Wrong value: expected `Text` but got something different.
2. `parseText` function failed to parse due to some reason.

I think better solution here is the following:

1. Change type of `parseText` argument from `Text -> Maybe a` to `Text -> Either Text a` where `Left` constructor will hold error messages. In most cases people should pass `readEither` function as an argument for `_TextBy`.
2. Change type of `matchValue` inside `mkAnyValueBiMap` from `forall t . Value t -> Maybe a` to `forall t . Value t -> Either TomlBiMapError a`

So this will require you to change types of all functions like:

```haskell
matchText :: Value t -> Maybe Text
```

to

```haskell
matchText :: Value t -> Either MatchError Text
```

Where `MatchError` looks like `TypeMismatchError`:

* https://github.com/kowainik/tomland/blob/4a47b14ac1b8122a6e4658a549da1a27492f7e85/src/Toml/Type/Value.hs#L209-L213

But instead of storing types, it should store type of expected value and actual value. And, I guess, instead of having two fields in `WrongValue` constructor, you can have just one field called `MatchError`.